BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -115,22 +115,41 @@ def new_personal #:nodoc:
     @provider         = UserkeyFlatDirSshDataProvider.new( :user_id   => current_user.id,
                                                            :group_id  => provider_group_id,
                                                            :online    => true,
-                                                           :read_only => false
+                                                           :read_only => false,
                                                          )
     @groups           = current_user.assignable_groups
   end
 
-  # create by normal user,  only UserkeyFlatDirSshDataProvider
+  # Create by normal user, only UserkeyFlatDirSshDataProvider
+  # S3FlatDataProvider, S3MultiLevelDataProvider
   def create_personal
-    normal_params = params.require_as_params(:data_provider)
-                          .permit(:name, :description, :group_id,
-                                  :remote_user, :remote_host,
-                                  :remote_port, :remote_dir
-                                 )
-    group_id = normal_params[:group_id]
-    current_user.assignable_group_ids.find(group_id) # ensure assignable, not sure need check visibility etc more
-    @provider         = UserkeyFlatDirSshDataProvider.new(normal_params)
-    @provider.user_id = current_user.id # prevent creation of dp on behalf of other users
+    dataprovider_params = params[:data_provider]
+
+    # Check if the type is an authorized class
+    dp_type             = (dataprovider_params[:type] || "UserkeyFlatDirSshDataProvider").constantize
+    authorized_type     = [UserkeyFlatDirSshDataProvider, S3FlatDataProvider, S3MultiLevelDataProvider]
+
+    if ! authorized_type.include?(dp_type)
```

Comment:
Make a check on the type here means that when the type is wrong, and the form is rendered again, the user will lose all the values that they entered in the form. It's not much of a problem because of course the form normally doesn't allow the user to change the type to something wrong, but it's something to think about. The normal workflow for a create action in rails is 1) create the object 2) validate 3) re-render the form if validation fails 4) save the object in database if validation succeeds